onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -0,0 +1,188 @@
+    }
+  }
+
+  const uint8_t* update_data = reinterpret_cast<const uint8_t*>(updates_input->DataRaw());
+  for (int64_t index = 0; index < num_indices;) {
+    const Tin axis_idx = indices_data[index];
+
+    // Compute the offset
+    size_t dst_offset = 0;
+    for (size_t i = 0; i < num_dims; ++i) {
+      if (i == size_t(axis)) {
+        dst_offset += axis_idx * out_weights[i];
+      } else {
+        dst_offset += dim_counters[i] * out_weights[i];
+      }
+    }
+
+    const size_t dst_offset_bytes = dst_offset * element_bytes;
+    assert(dst_offset_bytes < total_input_bytes);
+    if (is_string_type) {
+      reinterpret_cast<std::string*>(dst_base)[dst_offset] =
+          reinterpret_cast<const std::string*>(update_data)[index];
+    } else {
+      // Copy the element
+      // TODO: Replace memcpy
+      auto src_offset_bytes = index * element_bytes;
+      memcpy(dst_base + dst_offset_bytes, update_data + src_offset_bytes, element_bytes);
+    }
+
+    if (++index == num_indices) {
+      break;
+    }
+    // Increment counters
+    for (int64_t i = int64_t(num_dims - 1); i >= 0; --i) {
+      auto v = ++dim_counters[i];
+      assert(v <= upd_shape[i]);
+      if (v < upd_shape[i]) {
+        // No carry, done
+        break;
+      }
+      // No carry for the most significant dim
+      assert(i > 0);
+      dim_counters[i] = 0;
+    }
+  }
+  return Status::OK();
+}
+
+Status Scatter::Compute(OpKernelContext* context) const {
+  const auto* data_input = context->Input<Tensor>(0);
+  const auto& input_data_shape = data_input->Shape();
+  const auto axis = HandleNegativeAxis(axis_, input_data_shape.NumDimensions());
+
+  const auto* indices_input = context->Input<Tensor>(1);
+  const auto* updates_input = context->Input<Tensor>(2);
+
+  if (data_input->DataType() != updates_input->DataType()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "data type is different from updates type");
+  }
+
+  auto& indices_dims = indices_input->Shape().GetDims();
+  auto& updates_dims = updates_input->Shape().GetDims();
+  if (indices_dims.size() != updates_dims.size()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Indices and updates must have the same rank");
+  }
+
+  for (size_t i = 0; i < indices_dims.size(); ++i) {
+    if (indices_dims[i] != updates_dims[i]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices vs updates dimensions differs at position=", i,
+                             " ", indices_dims[i], " vs ", updates_dims[i]);
+    }
+  }
+
+  // According to the spec the rank of ind/upd shall be the same as input(data)
+  // and we also want to make sure that the dimensions of the of the ind/upd do not
+  // exceed that of the input
+  auto& input_dims = input_data_shape.GetDims();
+  if (input_dims.size() != indices_dims.size()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices/updates must have the same rank as input ind rank=",
```

Comment:
>rank as input ind rank [](start = 94, length = 22)

Doesn't the shape inferencing take care of this?

If not, "rank as input ind rank" is a bit cryptic. As this is an error message for a user, maybe "... same rank as the input. Indicies rank=", ..., ". Input rank=",...

Reply:
Shape inferencing doesn't guarantee to catch all errors, since in some cases the ranks may be statically unknown.

Reply:
Makes sense. Error message could still be improved for user friendliness.